src/Action.ts

```diff
@@ -1,141 +1,157 @@
-
-        return draftRelease?.id
-    }
-
-    private async createRelease(): Promise<CreateReleaseResponse> {
-        return await this.releases.create(
-            this.inputs.tag,
-            this.inputs.createdReleaseBody,
-            this.inputs.commit,
-            this.inputs.discussionCategory,
-            this.inputs.createdDraft,
-            this.inputs.generateReleaseNotes,
-            this.inputs.makeLatest,
-            this.inputs.createdReleaseName,
-            this.inputs.createdPrerelease
-        )
-    }
+	private inputs: Inputs;
+	private outputs: Outputs;
+	private releases: Releases;
+	private uploader: ArtifactUploader;
+	private artifactDestroyer: ArtifactDestroyer;
+	private skipper: ActionSkipper;
+
+	private releaseValidator: ReleaseValidator;
+
+	constructor(
+		inputs: Inputs,
+		outputs: Outputs,
+		releases: Releases,
+		uploader: ArtifactUploader,
+		artifactDestroyer: ArtifactDestroyer,
+		skipper: ActionSkipper,
+	) {
+		this.inputs = inputs;
+		this.outputs = outputs;
+		this.releases = releases;
+		this.uploader = uploader;
+		this.artifactDestroyer = artifactDestroyer;
+		this.skipper = skipper;
+		this.releaseValidator = new ReleaseValidator(inputs.updateOnlyUnreleased);
+	}
+
+	async perform() {
+		if (await this.skipper.shouldSkip()) {
+			core.notice(
+				"Skipping action, release already exists and skipIfReleaseExists is enabled.",
+			);
+			return;
+		}
+
+		const releaseResponse = await this.createOrUpdateRelease();
+		const releaseData = releaseResponse.data;
+		const releaseId = releaseData.id;
+		const uploadUrl = releaseData.upload_url;
+
+		if (this.inputs.removeArtifacts) {
+			await this.artifactDestroyer.destroyArtifacts(releaseId);
+		}
+
+		const artifacts = this.inputs.artifacts;
+		if (artifacts.length > 0) {
+			await this.uploader.uploadArtifacts(artifacts, releaseId, uploadUrl);
+		}
+
+		this.outputs.applyReleaseData(releaseData);
+	}
+
+	private async createOrUpdateRelease(): Promise<CreateOrUpdateReleaseResponse> {
+		if (this.inputs.allowUpdates) {
+			let getResponse: ReleaseByTagResponse;
+			try {
+				getResponse = await this.releases.getByTag(this.inputs.tag);
+			} catch (error: any) {
+				return await this.checkForMissingReleaseError(error);
+			}
+
+			// Fail if this isn't an unreleased release & updateOnlyUnreleased is enabled.
+			this.releaseValidator.validateReleaseUpdate(getResponse.data);
+
+			return await this.updateRelease(getResponse.data.id);
+		}
+
+		return await this.createRelease();
+	}
+
+	private async checkForMissingReleaseError(
+		error: Error,
+	): Promise<CreateOrUpdateReleaseResponse> {
+		if (Action.noPublishedRelease(error)) {
+			return await this.updateDraftOrCreateRelease();
+		}
+
+		throw error;
+	}
+
+	private async updateRelease(id: number): Promise<UpdateReleaseResponse> {
+		const response = await this.releases.generateReleaseNotes(this.inputs.tag);
```

Comment:
Could we check the inputs value here (`generateReleaseNotes`) before we hit this endpoint?